Keep decoded image data for background tabs in volatile memory

heycam opened the pull request:
#### 07a13e1a730be11a72625ddb74921a06e6c0c506
<pre>
Keep decoded image data for background tabs in volatile memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=254013">https://bugs.webkit.org/show_bug.cgi?id=254013</a>
rdar://106794310

Reviewed by NOBODY (OOPS!).

This patch introduces a cache of decoded image data that uses volatile
memory.

Currently, with the GPU process enabled, any time we draw a NativeImage,
the Web process decodes the image by drawing it into a ShareableBitmap,
then sends a handle to that ShareableBitmap to the GPU process for
drawing.  RemoteResourceCacheProxy takes that ShareableBitmap, wraps it
in a platform image, then sets it as the contents of the NativeImage.
This serves two purposes:

1. It allows us to re-use the same ShareableBitmap the next time we draw
   the same image.
2. It releases any platform cached data that might be lying behind the
   platform image that the NativeImage previously had as its contents.

The ShareableBitmap memory is non-volatile, which means that it
contributes to the memory footprint, and we won&apos;t release that memory
until the NativeImage is destroyed. For most images, this will be when
the object holding on to the CachedImage is destroyed (e.g. when an
&lt;img&gt; element is removed from the document), or when we are under memory
pressure and we explicitly discard decoded data from the MemoryCache.

To reduce the memory footprint, we use volatile memory for our decoded
image data in background tabs, where we know we&apos;re not likely to paint
them soon. When we next need to draw an image (e.g. when we switch back
to the tab), as long as we&apos;re not under memory pressure, then the system
will have preserved the volatile memory, and we can make it non-volatile
again in order to share it to the GPU process for drawing. If the system
has discarded the memory, we will re-decode the image.

A number of new classes are introduced:

* WebCore::DecodedImage: An abstract class that represents the source of
  image data for a NativeImage, which now holds one of these rather than
  a PlatformImagePtr. The DecodedImage interface has two main functions:

    - platformImage(), which provides the PlatformImagePtr
    - createCachedDecodedImage(), which can create an object to hold the
      volatile cached decoded image data

* WebCore::OwnedDecodedImage: A concrete DecodedImage that simply holds
  a PlatformImagePtr. A newly decoded image is created as a NativeObject
  holding an OwnedDecodedImage.

* WebKit::CacheableDecodedImage: A concrete DecodedImage that holds a
  ShareableBitmap and a PlatformImagePtr that wraps it.
  RemoteResourceCacheProxy::recordNativeImageUse creates one of these
  and sets it on the NativeImage. CacheableDecodedImage allows us to
  get an existing ShareableBitmap back from a NativeImage.

* WebCore::CachedDecodedImage: An abstract class that represents the
  decoded image data of a NativeImage in some cached form. It has one
  main function:

    - createNativeImage, which takes the contents of the
      CachedDecodedImage and creates a new NativeImage from it, if
      possible

* WebKit::PurgeableCachedDecodedImage: A concrete CachedDecodedImage
  that stores its data in volatile memory.

An example of how they are used, without being under memory pressure:

1. While painting a layer, we draw a BitmapImage&apos;s NativeImage to the
   recording GraphicsContext.
2. In RemoteResourceCacheProxy::recordNativeImageUse, we find there is
   no existing cached NativeImage, so we need to produce a
   ShareableBitmap to share with the GPU process.
3. We create a new ShareableBitmap, draw the NativeImage into it, wrap
   it in a new CacheableDecodedImage, and set the CacheableDecodedImage
   as the contents of the NativeImage.
4. Later, we switch tabs so that this current tab is now in the
   background. If WebProcess::pageWillLeaveWindow, we start
   the PurgeableCachedDecodedImageTimer.
5. When the timer fires shortly after, it tells the
   RemoteResourceCacheProxy to release all image resources, and then
   tells the MemoryCache to create a CachedDecodedImage for each
   NativeImage that can create one, and release the NativeImage.
6. Later, we switch back to this tab and paint the BitmapImage that
   originally contained the NativeImage. Upon finding that the
   NativeImage does not have a valid image frame, BitmapImage tries
   recreating it from the CachedDecodedImage stored on the ImageFrame.
   This is done in CacheableDecodedImage::createNativeImage, which tries
   to switch the memory back to non-volatile, and if successful,
   creates a new CacheableDecodedImage wrapping the memory, and a new
   NativeImage to wrap the CacheableDecodedImage.
7. In RemoteResourceCacheProxy::recordNativeImageUse, we find the new
   NativeImage is not in the resource cache, but we do find that it
   already has a CacheableDecodedImage. We grab the ShareableBitmap from
   it to avoid creating a new one, and send it to the GPU process.

If memory pressure does hit after the image data has been made volatile,
then BitmapImage will re-decode the image, just as if the MemoryCache
had explicitly discarded the decoded image data.

In PurgeableCachedDecodedImage, we don&apos;t take the existing memory
ShareableBitmap memory allocation to be volatile, and instead we create
a new SharedMemory, copy the existing (non-volatile) data into it, then
mark it as volatile. This is because we can&apos;t be sure that the
ShareableBitmap is still not being used by the GPU process. We could
have some way for the GPU process to signal back to the Web process that
its ShareableBitmap has been unmapped and destroyed, but it would be
tricky to do that from the ShareableBitmap destructor.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::destroyDecodedDataForAllImages):
(WebCore::MemoryCache::destroyAndCacheDecodedDataForAllImages):
* Source/WebCore/loader/cache/MemoryCache.h:
* Source/WebCore/platform/VolatilityTypes.h: Added.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::destroyAndCacheDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::internalStartAnimation):
(WebCore::BitmapImage::decode):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/CachedDecodedImage.cpp: Added.
* Source/WebCore/platform/graphics/CachedDecodedImage.h: Added.
* Source/WebCore/platform/graphics/DecodedImage.cpp: Added.
(WebCore::DecodedImage::DecodedImage):
* Source/WebCore/platform/graphics/DecodedImage.h: Added.
(WebCore::DecodedImage::type const):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::destroyAndCacheDecodedData):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::clearImageAndCacheDecodedData):
(WebCore::ImageFrame::clearImage):
(WebCore::ImageFrame::clear):
(WebCore::ImageFrame::hasCachedDecodedImage const):
(WebCore::ImageFrame::hasCachedDecodedImageCompatibleWithOptions const):
(WebCore::ImageFrame::operator=): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::hasCachedDecodedImage):
(WebCore::ImageFrame::ImageFrame): Deleted.
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::destroyAndCacheDecodedData):
(WebCore::ImageSource::cacheNativeImageAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndex):
(WebCore::ImageSource::createFrameImageFromCachedDecodedImageAtIndex):
(WebCore::ImageSource::setFrameImageFromCachedDecodedImageAtIndex):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
(WebCore::ImageSource::frameHasDecodedNativeImageCompatibleWithOptionsAtIndex):
(WebCore::ImageSource::frameHasCachedDecodedImageCompatibleWithOptionsAtIndex):
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::setPlatformImage): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/OwnedDecodedImage.cpp: Added.
(WebCore::OwnedDecodedImage::create):
(WebCore::OwnedDecodedImage::OwnedDecodedImage):
(WebCore::OwnedDecodedImage::platformImage const):
(WebCore::OwnedDecodedImage::createCachedDecodedImage const):
* Source/WebCore/platform/graphics/OwnedDecodedImage.h: Added.
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::size const):
(WebCore::NativeImage::hasAlpha const):
(WebCore::NativeImage::singlePixelSolidColor const):
(WebCore::NativeImage::colorSpace const):
(WebCore::NativeImage::clearSubimages):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemory::setVolatile):
(WebKit::SharedMemory::setNonVolatile):
* Source/WebKit/Shared/CacheableDecodedImage.cpp: Added.
(WebKit::CacheableDecodedImage::create):
(WebKit::CacheableDecodedImage::CacheableDecodedImage):
(WebKit::CacheableDecodedImage::platformImage const):
(WebKit::CacheableDecodedImage::createCachedDecodedImage const):
* Source/WebKit/Shared/CacheableDecodedImage.h: Added.
(isType):
* Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp: Added.
(WebKit::PurgeableCachedDecodedImageTimer::singleton):
(WebKit::PurgeableCachedDecodedImageTimer::start):
(WebKit::PurgeableCachedDecodedImageTimer::defer):
(WebKit::PurgeableCachedDecodedImageTimer::stop):
(WebKit::PurgeableCachedDecodedImageTimer::timerFired):
(WebKit::PurgeableCachedDecodedImage::create):
(WebKit::PurgeableCachedDecodedImage::createVolatileCopyOfBitmap):
(WebKit::PurgeableCachedDecodedImage::PurgeableCachedDecodedImage):
(WebKit::PurgeableCachedDecodedImage::createNativeImage):
(WebKit::PurgeableCachedDecodedImage::startCleanup):
(WebKit::PurgeableCachedDecodedImage::stopCleanup):
* Source/WebKit/Shared/PurgeableCachedDecodedImage.h: Added.
* Source/WebKit/Shared/ShareableBitmap.h:
(WebKit::ShareableBitmap::configuration const):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::getExistingShareableBitmapFromNativeImage):
(WebKit::createShareableBitmapFromNativeImage):
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::pageDidEnterWindow):
(WebKit::WebProcess::pageWillLeaveWindow):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07a13e1a730be11a72625ddb74921a06e6c0c506

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/851 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/874 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/907 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1195 "Failed to compile WebKit") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/752 "Failed to compile WebKit") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/840 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/959 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/1195 "Failed to compile WebKit") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/802 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1125 "Failed to compile WebKit") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/859 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/800 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/1125 "Failed to compile WebKit") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/748 "Found 1 jsc stress test failure: wasm.yaml/wasm/stress/simple-inline-stacktrace-with-catch.js.wasm-agressive-inline") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/770 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/823 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/1125 "Failed to compile WebKit") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/836 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/775 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/876 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/808 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/185 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/821 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/879 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/179 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -56,6 +56,7 @@ class SharedBuffer;
 namespace WebKit {
 
 enum class MemoryLedger { None, Default, Network, Media, Graphics, Neural };
+enum class WasPurged : bool { No, Yes };
```
Maybe reuse or share with
```
enum class SetNonVolatileResult : uint8_t {
    Valid,
    Empty
};

enum class VolatilityState : uint8_t {
    NonVolatile,
    Volatile
};
```

heycam committed:
Keep decoded image data for background tabs in volatile memory
https://bugs.webkit.org/show_bug.cgi?id=254013
rdar://106794310

Reviewed by NOBODY (OOPS!).

This patch introduces a cache of decoded image data that uses volatile
memory.

Currently, with the GPU process enabled, any time we draw a NativeImage,
the Web process decodes the image by drawing it into a ShareableBitmap,
then sends a handle to that ShareableBitmap to the GPU process for
drawing.  RemoteResourceCacheProxy takes that ShareableBitmap, wraps it
in a platform image, then sets it as the contents of the NativeImage.
This serves two purposes:

1. It allows us to re-use the same ShareableBitmap the next time we draw
   the same image.
2. It releases any platform cached data that might be lying behind the
   platform image that the NativeImage previously had as its contents.

The ShareableBitmap memory is non-volatile, which means that it
contributes to the memory footprint, and we won't release that memory
until the NativeImage is destroyed. For most images, this will be when
the object holding on to the CachedImage is destroyed (e.g. when an
<img> element is removed from the document), or when we are under memory
pressure and we explicitly discard decoded data from the MemoryCache.

To reduce the memory footprint, we use volatile memory for our decoded
image data in background tabs, where we know we're not likely to paint
them soon. When we next need to draw an image (e.g. when we switch back
to the tab), as long as we're not under memory pressure, then the system
will have preserved the volatile memory, and we can make it non-volatile
again in order to share it to the GPU process for drawing. If the system
has discarded the memory, we will re-decode the image.

A number of new classes are introduced:

* WebCore::DecodedImage: An abstract class that represents the source of
  image data for a NativeImage, which now holds one of these rather than
  a PlatformImagePtr. The DecodedImage interface has two main functions:

    - platformImage(), which provides the PlatformImagePtr
    - createCachedDecodedImage(), which can create an object to hold the
      volatile cached decoded image data

* WebCore::OwnedDecodedImage: A concrete DecodedImage that simply holds
  a PlatformImagePtr. A newly decoded image is created as a NativeObject
  holding an OwnedDecodedImage.

* WebKit::CacheableDecodedImage: A concrete DecodedImage that holds a
  ShareableBitmap and a PlatformImagePtr that wraps it.
  RemoteResourceCacheProxy::recordNativeImageUse creates one of these
  and sets it on the NativeImage. CacheableDecodedImage allows us to
  get an existing ShareableBitmap back from a NativeImage.

* WebCore::CachedDecodedImage: An abstract class that represents the
  decoded image data of a NativeImage in some cached form. It has one
  main function:

    - createNativeImage, which takes the contents of the
      CachedDecodedImage and creates a new NativeImage from it, if
      possible

* WebKit::PurgeableCachedDecodedImage: A concrete CachedDecodedImage
  that stores its data in volatile memory.

An example of how they are used, without being under memory pressure:

1. While painting a layer, we draw a BitmapImage's NativeImage to the
   recording GraphicsContext.
2. In RemoteResourceCacheProxy::recordNativeImageUse, we find there is
   no existing cached NativeImage, so we need to produce a
   ShareableBitmap to share with the GPU process.
3. We create a new ShareableBitmap, draw the NativeImage into it, wrap
   it in a new CacheableDecodedImage, and set the CacheableDecodedImage
   as the contents of the NativeImage.
4. Later, we switch tabs so that this current tab is now in the
   background. If WebProcess::pageWillLeaveWindow, we start
   the PurgeableCachedDecodedImageTimer.
5. When the timer fires shortly after, it tells the
   RemoteResourceCacheProxy to release all image resources, and then
   tells the MemoryCache to create a CachedDecodedImage for each
   NativeImage that can create one, and release the NativeImage.
6. Later, we switch back to this tab and paint the BitmapImage that
   originally contained the NativeImage. Upon finding that the
   NativeImage does not have a valid image frame, BitmapImage tries
   recreating it from the CachedDecodedImage stored on the ImageFrame.
   This is done in CacheableDecodedImage::createNativeImage, which tries
   to switch the memory back to non-volatile, and if successful,
   creates a new CacheableDecodedImage wrapping the memory, and a new
   NativeImage to wrap the CacheableDecodedImage.
7. In RemoteResourceCacheProxy::recordNativeImageUse, we find the new
   NativeImage is not in the resource cache, but we do find that it
   already has a CacheableDecodedImage. We grab the ShareableBitmap from
   it to avoid creating a new one, and send it to the GPU process.

If memory pressure does hit after the image data has been made volatile,
then BitmapImage will re-decode the image, just as if the MemoryCache
had explicitly discarded the decoded image data.

In PurgeableCachedDecodedImage, we don't take the existing memory
ShareableBitmap memory allocation to be volatile, and instead we create
a new SharedMemory, copy the existing (non-volatile) data into it, then
mark it as volatile. This is because we can't be sure that the
ShareableBitmap is still not being used by the GPU process. We could
have some way for the GPU process to signal back to the Web process that
its ShareableBitmap has been unmapped and destroyed, but it would be
tricky to do that from the ShareableBitmap destructor.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::destroyDecodedDataForAllImages):
(WebCore::MemoryCache::destroyAndCacheDecodedDataForAllImages):
* Source/WebCore/loader/cache/MemoryCache.h:
* Source/WebCore/platform/VolatilityTypes.h: Added.
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::destroyAndCacheDecodedData):
(WebCore::BitmapImage::draw):
(WebCore::BitmapImage::internalStartAnimation):
(WebCore::BitmapImage::decode):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/CachedDecodedImage.cpp: Added.
* Source/WebCore/platform/graphics/CachedDecodedImage.h: Added.
* Source/WebCore/platform/graphics/DecodedImage.cpp: Added.
(WebCore::DecodedImage::DecodedImage):
* Source/WebCore/platform/graphics/DecodedImage.h: Added.
(WebCore::DecodedImage::type const):
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::destroyAndCacheDecodedData):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/ImageFrame.cpp:
(WebCore::ImageFrame::ImageFrame):
(WebCore::ImageFrame::clearImageAndCacheDecodedData):
(WebCore::ImageFrame::clearImage):
(WebCore::ImageFrame::clear):
(WebCore::ImageFrame::hasCachedDecodedImage const):
(WebCore::ImageFrame::hasCachedDecodedImageCompatibleWithOptions const):
(WebCore::ImageFrame::operator=): Deleted.
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::hasCachedDecodedImage):
(WebCore::ImageFrame::ImageFrame): Deleted.
* Source/WebCore/platform/graphics/ImageSource.cpp:
(WebCore::ImageSource::destroyAndCacheDecodedData):
(WebCore::ImageSource::cacheNativeImageAtIndex):
(WebCore::ImageSource::cachePlatformImageAtIndex):
(WebCore::ImageSource::createFrameImageFromCachedDecodedImageAtIndex):
(WebCore::ImageSource::setFrameImageFromCachedDecodedImageAtIndex):
(WebCore::ImageSource::frameAtIndexCacheIfNeeded):
(WebCore::ImageSource::frameHasDecodedNativeImageCompatibleWithOptionsAtIndex):
(WebCore::ImageSource::frameHasCachedDecodedImageCompatibleWithOptionsAtIndex):
* Source/WebCore/platform/graphics/ImageSource.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::create):
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::setPlatformImage): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/OwnedDecodedImage.cpp: Added.
(WebCore::OwnedDecodedImage::create):
(WebCore::OwnedDecodedImage::OwnedDecodedImage):
(WebCore::OwnedDecodedImage::platformImage const):
(WebCore::OwnedDecodedImage::createCachedDecodedImage const):
* Source/WebCore/platform/graphics/OwnedDecodedImage.h: Added.
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::size const):
(WebCore::NativeImage::hasAlpha const):
(WebCore::NativeImage::singlePixelSolidColor const):
(WebCore::NativeImage::colorSpace const):
(WebCore::NativeImage::clearSubimages):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemory::setVolatile):
(WebKit::SharedMemory::setNonVolatile):
* Source/WebKit/Shared/CacheableDecodedImage.cpp: Added.
(WebKit::CacheableDecodedImage::create):
(WebKit::CacheableDecodedImage::CacheableDecodedImage):
(WebKit::CacheableDecodedImage::platformImage const):
(WebKit::CacheableDecodedImage::createCachedDecodedImage const):
* Source/WebKit/Shared/CacheableDecodedImage.h: Added.
(isType):
* Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp: Added.
(WebKit::PurgeableCachedDecodedImageTimer::singleton):
(WebKit::PurgeableCachedDecodedImageTimer::start):
(WebKit::PurgeableCachedDecodedImageTimer::defer):
(WebKit::PurgeableCachedDecodedImageTimer::stop):
(WebKit::PurgeableCachedDecodedImageTimer::timerFired):
(WebKit::PurgeableCachedDecodedImage::create):
(WebKit::PurgeableCachedDecodedImage::createVolatileCopyOfBitmap):
(WebKit::PurgeableCachedDecodedImage::PurgeableCachedDecodedImage):
(WebKit::PurgeableCachedDecodedImage::createNativeImage):
(WebKit::PurgeableCachedDecodedImage::startCleanup):
(WebKit::PurgeableCachedDecodedImage::stopCleanup):
* Source/WebKit/Shared/PurgeableCachedDecodedImage.h: Added.
* Source/WebKit/Shared/ShareableBitmap.h:
(WebKit::ShareableBitmap::configuration const):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::getExistingShareableBitmapFromNativeImage):
(WebKit::createShareableBitmapFromNativeImage):
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::pageDidEnterWindow):
(WebKit::WebProcess::pageWillLeaveWindow):

shallawa commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -882,3 +882,7 @@
 #define ENABLE_WEBGL_IN_WORKERS 1
 #endif
 
+#if !defined(ENABLE_PURGEABLE_DECODED_IMAGE_DATA_CACHE)
+#define ENABLE_PURGEABLE_DECODED_IMAGE_DATA_CACHE 1
+#endif
+
```
Why this code is enabled for Cocoa platforms only?

heycam replied:
Because I haven't implemented the other platform backends for SharedMemory::setVolatile/setNonVolatile. It's probably doable but I'm less familiar with the other platforms.

shallawa commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -239,7 +239,7 @@ void MemoryCache::forEachSessionResource(PAL::SessionID sessionID, const Functio
 void MemoryCache::destroyDecodedDataForAllImages()
 {
-    MemoryCache::singleton().forEachResource([](CachedResource& resource) {
+    forEachResource([](CachedResource& resource) {
```
Why do not we make `destroyDecodedDataForAllImages()` be static and make the callers call `MemoryCache::destroyDecodedDataForAllImages()` instead of calling `MemoryCache::singleton().destroyDecodedDataForAllImages()`?

heycam replied:
Can do that.

shallawa commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -248,6 +248,17 @@ void MemoryCache::destroyDecodedDataForAllImages()
     });
 }
 
+void MemoryCache::destroyAndCacheDecodedDataForAllImages()
```
This name is a little bit confusing. Does it destroy and cache decoded data at the same time? I think we can just mention the intention of this function and hide the details. Maybe something like `markDecodedDataVolatileForAllImages()` or `makeDecodedDataVolatileForAllImages()` would be clearer?

Also it can be also static I think.

heycam replied:
I thought it was preferable to avoid mentioning the concept of volatility in WebCore code. But if it's OK, I can go with your name suggestions.

shallawa commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -248,6 +248,17 @@ void MemoryCache::destroyDecodedDataForAllImages()
+            return;
+
+        if (auto image = downcast<CachedImage>(resource).image())
+            image->destroyAndCacheDecodedData();
```
Ditto.

shallawa commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -248,6 +248,17 @@ void MemoryCache::destroyDecodedDataForAllImages()
+void MemoryCache::destroyAndCacheDecodedDataForAllImages()
+{
+    forEachResource([](CachedResource& resource) {
+        if (!resource.isImage())
+            return;
+
+        if (auto image = downcast<CachedImage>(resource).image())
```
if (auto image = dynamicDowncast<CachedImage>(resource).image())

heycam replied:
Would still need to null check the dynamicDowncast result.

shallawa commented on Source/WebCore/platform/VolatilityTypes.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace WebCore {
+
+enum class SetNonVolatileResult : uint8_t { Valid, Empty };
```
Not new. Why this one has the name `SetNonVolatileResult` even it is the return value of 
```
SetNonVolatileResult IOSurface::setVolatile(bool isVolatile)
```

heycam replied:
Not added, just moved from ImageBufferBackend.h so I had a more convenient place to include it from.

heycam replied:
(I misinterpreted your "not new". I thought you meant "this type is not new, i.e. it's duplicating an existing type", but instead you meant "this is a comment on existing code".)

shallawa commented on Source/WebCore/platform/VolatilityTypes.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+enum class SetNonVolatileResult : uint8_t { Valid, Empty };
+enum class VolatilityState : uint8_t { NonVolatile, Volatile };
```
Not new. But should not VolatilityState be named `MemoryState { NonVolatile, Volatile }` since this is what it describes?

heycam replied:
Actually I didn't end up using this type (an earlier iteration did use it), so I can leave it where it is in ImageBufferBackend.h.

shallawa commented on Source/WebCore/platform/graphics/BitmapImage.cpp:
```diff
@@ -605,8 +622,10 @@ void BitmapImage::decode(Function<void()>&& callback)
     else if (!frameIsBeingDecoded) {
-        m_source->requestFrameAsyncDecodingAtIndex(m_currentFrame, m_currentSubsamplingLevel, std::optional<IntSize>());
-        m_currentFrameDecodingStatus = DecodingStatus::Decoding;
+        if (!m_source->setFrameImageFromCachedDecodedImageAtIndex(m_currentFrame, m_currentSubsamplingLevel, DecodingMode::Asynchronous)) {
+            m_source->requestFrameAsyncDecodingAtIndex(m_currentFrame, m_currentSubsamplingLevel);
+            m_currentFrameDecodingStatus = DecodingStatus::Decoding;
+        }
```
This code has been repeated almost the same four times. Can't we make 

```
DecodingStatus  ImageSource:: requestFrameAsyncDecodingAtIndex(...)
{
    if (setFrameImageFromCachedDecodedImageAtIndex(index, subsamplingLevel, decodingOptions)
        return DecodingStatus::Complete;
    ...
}
```

heycam replied:
Yes, we can put it down in requestFrameAsyncDecodingAtIndex.

shallawa commented on Source/WebCore/platform/graphics/BitmapImage.h:
```diff
@@ -101,6 +101,7 @@ class BitmapImage final : public Image {
 
     bool frameHasFullSizeNativeImageAtIndex(size_t index, SubsamplingLevel subsamplingLevel) { return m_source->frameHasFullSizeNativeImageAtIndex(index, subsamplingLevel); }
     bool frameHasDecodedNativeImageCompatibleWithOptionsAtIndex(size_t index, const std::optional<SubsamplingLevel>& subsamplingLevel, const DecodingOptions& decodingOptions) { return m_source->frameHasDecodedNativeImageCompatibleWithOptionsAtIndex(index, subsamplingLevel, decodingOptions); }
+    bool frameHasCachedDecodedImageCompatibleWithOptionsAtIndex(size_t index, const std::optional<SubsamplingLevel>& subsamplingLevel, const DecodingOptions& decodingOptions) { return m_source->frameHasCachedDecodedImageCompatibleWithOptionsAtIndex(index, subsamplingLevel, decodingOptions); }
```
I do not think this function is ever called.

heycam replied:
Yes, that's a remnant from an earlier iteration where I was calling this first before calling setFrameImageFromCachedDecodedImageAtIndex.

shallawa commented on Source/WebCore/platform/graphics/CachedDecodedImage.cpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace WebCore {
+
+CachedDecodedImage::CachedDecodedImage() = default;
+CachedDecodedImage::~CachedDecodedImage() = default;
```
why are not these just inlined in the header filer?

heycam replied:
Just had a slight preference not to have an inline virtual function. Probably not a big deal since hopefully the linker will deduplicate the different instances that get generated for the vtables of descendant classes.

shallawa commented on Source/WebKit/Shared/CacheableDecodedImage.h:
```diff
@@ -0,0 +1,56 @@
+
+    ~CacheableDecodedImage() final;
+    WebCore::PlatformImagePtr platformImage() const final;
+    std::unique_ptr<CachedDecodedImage> createCachedDecodedImage() const final;
```
Here is some weird structure a `CacheableDecodedImage` creates a `CachedDecodedImage`. These  names make it  difficult to remember what exactly they mean.

heycam replied:
Yeah, I also did not really like the names...

shallawa commented on Source/WebCore/platform/graphics/CachedDecodedImage.h:
```diff
@@ -0,0 +1,41 @@
+
+class NativeImage;
+
+class CachedDecodedImage {
```
Can't `CachedDecodedImage` be renamed `DecodedImage`?
And `DecodedImage` can be renamed `DecodedImageSource`. 

It is really difficult to remember while reviewing this PR the difference between `DecodedImage`, `CachedDecodedImage` and the difference between `CachedDecodedImage` and `CacheableDecodedImage`.

heycam replied:
I think those names `DecodedImage` / `DecodedImageSource` are OK. I will use them.

heycam replied:
I guess it would be nice though if `DecodedImage` suggested somehow that it's a cache of decoded image data that's different from a `NativeImage` itself.

shallawa commented on Source/WebCore/platform/graphics/Image.h:
```diff
@@ -135,6 +135,7 @@ class Image : public RefCounted<Image> {
     virtual String accessibilityDescription() const { return String(); } // null string if unknown
 
     virtual void destroyDecodedData(bool /*destroyAll*/ = true) { }
+    virtual void destroyAndCacheDecodedData() { }
```
Maybe `setDecodedDataVolatile()` or `markDecodeDataPurgeable()`?

heycam replied:
As above, I was trying to avoid using the terminology "volatile" and "purgeable" since it seemed like these should be WebKit layer concepts.

shallawa commented on Source/WebCore/platform/graphics/DecodedImage.h:
```diff
@@ -0,0 +1,51 @@
+
+enum class DecodedImageType : uint8_t { Owned, Cacheable };
+
+class DecodedImage {
```
Can't `DecodedImage` be renamed `DecodedImageSource` or `DecodedImageProvider`?

shallawa commented on Source/WebCore/platform/graphics/DecodedImage.h:
```diff
@@ -0,0 +1,51 @@
+public:
+    WEBCORE_EXPORT virtual ~DecodedImage();
+    virtual PlatformImagePtr platformImage() const = 0;
+    virtual std::unique_ptr<CachedDecodedImage> createCachedDecodedImage() const = 0;
```
If this class is renamed `DecodedImageSource`, this method can be renamed 
```
   virtual std::unique_ptr<DecodedImage> createDecodedImage() const = 0;
```

shallawa commented on Source/WebCore/platform/graphics/DecodedImage.h:
```diff
@@ -0,0 +1,51 @@
+    WEBCORE_EXPORT explicit DecodedImage(DecodedImageType);
+
+private:
+    DecodedImageType m_type;
```
Why do we need this type and the inheritance at the same type?

shallawa commented on Source/WebCore/platform/graphics/ImageFrame.cpp:
```diff
@@ -45,22 +45,18 @@ const ImageFrame& ImageFrame::defaultFrame()
 }
 
-ImageFrame& ImageFrame::operator=(const ImageFrame& other)
+ImageFrame::ImageFrame(ImageFrame&& other)
```
Can't this be a default move constructor in the header file?
```
    ImageFrame(ImageFrame&&) = default;
```

heycam replied:
Yes it can.

shallawa commented on Source/WebCore/platform/graphics/ImageFrame.cpp:
```diff
@@ -75,24 +71,57 @@ DecodingStatus ImageFrame::decodingStatus() const
+{
     m_decodingOptions = DecodingOptions();
 
+    unsigned frameBytes = 0;
+
+    if (m_nativeImage) {
+        frameBytes = this->frameBytes();
+        m_nativeImage->clearSubimages();
+        m_nativeImage = nullptr;
+    }
+
+    m_cachedDecodedImage.reset();
+
```
Instead of this change, I think you can still make an early return if you add 
```
    m_cachedDecodedImage.reset();
```
before 
```
    if (!hasNativeImage())
        return 0;
```
in the original ImageFrame::clearImage().

shallawa commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -431,6 +448,30 @@ void ImageSource::stopAsyncDecodingQueue()
     LOG(Images, "ImageSource::%s - %p - url: %s [decoding has been stopped]", __FUNCTION__, this, sourceURL().string().utf8().data());
 }
 
+RefPtr<NativeImage> ImageSource::createFrameImageFromCachedDecodedImageAtIndex(size_t index, SubsamplingLevel subsamplingLevel, const DecodingOptions& decodingOptions)
```
This function is only called from `setFrameImageFromCachedDecodedImageAtIndex()`. Maybe moving its code to the caller will look better.

shallawa commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -431,6 +448,30 @@ void ImageSource::stopAsyncDecodingQueue()
+
+bool ImageSource::setFrameImageFromCachedDecodedImageAtIndex(size_t index, SubsamplingLevel subsamplingLevel, const DecodingOptions& decodingOptions)
+{
+    auto image = createFrameImageFromCachedDecodedImageAtIndex(index, subsamplingLevel, decodingOptions);
+    if (!image)
+        return false;
+
+    cacheNativeImageAtIndex(WTFMove(image), index, subsamplingLevel, decodingOptions);
```
It is a little bit strange that we get a NativeImage out of the ImageFrame (by calling createFrameImageFromCachedDecodedImageAtIndex ()) and then we set it back (by calling cacheNativeImageAtIndex()). Should not this be done in one function instead?

shallawa commented on Source/WebCore/platform/graphics/ImageSource.cpp:
```diff
@@ -291,14 +302,20 @@ void ImageSource::cachePlatformImageAtIndex(PlatformImagePtr&& platformImage, si
     // Move the new image to the cache.
-    frame.m_nativeImage = NativeImage::create(WTFMove(platformImage));
+    frame.m_nativeImage = WTFMove(nativeImage);
+    frame.m_cachedDecodedImage.reset();
```
I think we never hit the case where m_cachedDecodedImage is not nullptr. Maybe we need to `ASSERT(!frame.m_cachedDecodedImage);` instead.

shallawa commented on Source/WebCore/platform/graphics/ImageFrame.h:
```diff
@@ -88,13 +87,17 @@ class ImageFrame {
     DecodingStatus m_decodingStatus { DecodingStatus::Invalid };
     IntSize m_size;
 
     RefPtr<NativeImage> m_nativeImage;
+    std::unique_ptr<CachedDecodedImage> m_cachedDecodedImage;
```
Can an ImageFrame has valid NativeImage and CachedDecodedImage at the same time? If not should not these two members be a variant?

shallawa commented on Source/WebCore/platform/graphics/OwnedDecodedImage.cpp:
```diff
@@ -0,0 +1,58 @@
+    return m_platformImage;
+}
+
+std::unique_ptr<CachedDecodedImage> OwnedDecodedImage::createCachedDecodedImage() const
+{
+    return nullptr;
+}
```
I think DecodedImage::createCachedDecodedImage() should not be pure virtual. It should just return a nullptr and we can get rid of this one.

shallawa commented on Source/WebCore/platform/graphics/OwnedDecodedImage.cpp:
```diff
@@ -0,0 +1,58 @@
+
+OwnedDecodedImage::~OwnedDecodedImage() = default;
+
+PlatformImagePtr OwnedDecodedImage::platformImage() const
+{
+    return m_platformImage;
+}
```
This can just be inlined in the header file.

shallawa commented on Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
```diff
@@ -273,4 +273,32 @@ WTF::MachSendRight SharedMemory::createSendRight(Protection protection) const
+    int purgeableState = VM_PURGABLE_VOLATILE;
+    auto kr = mach_vm_purgable_control(mach_task_self(), toVMAddress(m_data), VM_PURGABLE_SET_STATE, &purgeableState);
+
+    RELEASE_LOG_ERROR_IF(kr != KERN_SUCCESS, VirtualMemory, "SharedMemory::setVolatile: Failed to set memory volatile. Error: %" PUBLIC_LOG_STRING " (%x)", mach_error_string(kr), kr);
```
Why do not we return whether the mach call has succeeded or not like what we do in `SharedMemory::setNonVolatile()`?

shallawa commented on Source/WebKit/Shared/CacheableDecodedImage.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace WebKit {
+
+class CacheableDecodedImage final : public WebCore::DecodedImage {
```
Should not this class be named `ShareableDecodedImage` since it holds a `ShareableBitmap`?

shallawa commented on Source/WebKit/Shared/CacheableDecodedImage.cpp:
```diff
@@ -0,0 +1,68 @@
+
+namespace WebKit {
+
+std::unique_ptr<CacheableDecodedImage> CacheableDecodedImage::create(RefPtr<ShareableBitmap>&& bitmap)
```
I think this function should take `Ref<ShareableBitmap>&&` since the caller makes the null check before calling it.

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp:
```diff
@@ -0,0 +1,153 @@
+    if (!bitmap)
+        return nullptr;
+
+    auto decodedImage = CacheableDecodedImage::create(WTFMove(bitmap));
```
I think you can pass `bitmap.releaseNonNull()` instead.

shallawa commented on Source/WebKit/Shared/CacheableDecodedImage.cpp:
```diff
@@ -0,0 +1,68 @@
+    if (!bitmap)
+        return nullptr;
+
+    auto platformImage = bitmap->createPlatformImage(DontCopyBackingStore, ShouldInterpolate::Yes);
+    if (!platformImage)
+        return nullptr;
```
Should not we lazily create `m_platformImage` in `CacheableDecodedImage::platformImage()`?

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.h:
```diff
@@ -0,0 +1,61 @@
+
+class SharedMemory;
+
+class PurgeableCachedDecodedImage final : public WebCore::CachedDecodedImage {
```
1. PurgeableCachedDecodedImage::create() is called form CacheableDecodedImage::createCachedDecodedImage()
2. CacheableDecodedImage::create() is called form PurgeableCachedDecodedImage::createNativeImage()
3. CacheableDecodedImage::create() is called form createShareableBitmapFromNativeImage() in RemoteResourceCacheProxy.cpp

So I think CacheableDecodedImage and PurgeableCachedDecodedImage should not be in WebKit/Shared folder. I think they should be in WebKit/WebProcess/GPU/graphics instead.

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp:
```diff
@@ -0,0 +1,153 @@
+    return NativeImage::create(makeUniqueRefFromNonNullUniquePtr(WTFMove(decodedImage)));
+}
+
+void PurgeableCachedDecodedImage::startCleanup()
+{
+    PurgeableCachedDecodedImageTimer::singleton().start();
+}
+
+void PurgeableCachedDecodedImage::stopCleanup()
+{
+    PurgeableCachedDecodedImageTimer::singleton().stop();
+}
```
Why do we need to include these static functions in this class? They release all the remote images not only a single one. So maybe these two functions can be static in `PurgeableCachedDecodedImageTimer` itself. But I think it should be renamed in this case since it will have the timer and the handlers as well.

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp:
```diff
@@ -0,0 +1,153 @@
+    if (!m_memory)
+        return nullptr;
+
+    if (m_memory->setNonVolatile() == WebCore::SetNonVolatileResult::Empty) {
```
I am not sure if this violates our code style or not. But I think we do not like hidden side effect like this one.
```
    auto result = m_memory->setNonVolatile();
    if (result == WebCore::SetNonVolatileResult::Empty)
```

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp:
```diff
@@ -0,0 +1,153 @@
+
+RefPtr<NativeImage> PurgeableCachedDecodedImage::createNativeImage()
+{
+    if (!m_memory)
+        return nullptr;
```
I think we should `ASSERT(m_memory);` here. If we find `m_memory` is null that means this function was called before and we tried to `m_memory->setNonVolatile()` but it was unsuccessful so we set m_memory to null. So what is the point to call this function again?

shallawa commented on Source/WebKit/Shared/PurgeableCachedDecodedImage.cpp:
```diff
@@ -0,0 +1,153 @@
+
+static constexpr Seconds cleanupDelay { 2_s };
+
+class PurgeableCachedDecodedImageTimer {
```
I think this class deserves its own source and header files since it woks on all images and not `PurgeableCachedDecodedImage` only. It calls even `MemoryCache::singleton().destroyAndCacheDecodedDataForAllImages();`

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -82,8 +83,19 @@ void RemoteResourceCacheProxy::forgetImageBuffer(RenderingResourceIdentifier ide
     ASSERT_UNUSED(success, success);
 }
 
+inline static RefPtr<ShareableBitmap> getExistingShareableBitmapFromNativeImage(NativeImage& image)
+{
+    if (!is<CacheableDecodedImage>(image.decodedImage()))
+        return nullptr;
+
+    return downcast<CacheableDecodedImage>(image.decodedImage()).bitmap();
+}
```
I think this can be a lambda inside `createShareableBitmapFromNativeImage()`.

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -82,8 +83,19 @@ void RemoteResourceCacheProxy::forgetImageBuffer(RenderingResourceIdentifier ide
+
 inline static RefPtr<ShareableBitmap> createShareableBitmapFromNativeImage(NativeImage& image)
 {
+    if (auto decodedImage = dynamicDowncast<CacheableDecodedImage>(image.decodedImage()))
+        return decodedImage->bitmap();
```
Is not this code the same as `getExistingShareableBitmapFromNativeImage()`?

shallawa commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -110,19 +122,23 @@ void RemoteResourceCacheProxy::recordNativeImageUse(NativeImage& image)
-    auto bitmap = createShareableBitmapFromNativeImage(image);
-    if (!bitmap)
-        return;
+    auto bitmap = getExistingShareableBitmapFromNativeImage(image);
```
I think we do not need to make this call. `createShareableBitmapFromNativeImage()` will do it anyway and will make an early return with `decodedImage->bitmap()` if it exists otherwise it will create a new ShareableBitmap.

shallawa commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1668,6 +1673,10 @@ void WebProcess::pageWillLeaveWindow(PageIdentifier pageID)
             m_nonVisibleProcessMemoryCleanupTimer.startOneShot(nonVisibleProcessMemoryCleanupDelay);
 #endif
+
+#if ENABLE(PURGEABLE_DECODED_IMAGE_DATA_CACHE)
+        PurgeableCachedDecodedImage::startCleanup();
+#endif
```
Why do we need to add a new timer for cleaning up the images? Can't we use one of the existing timers instead: `m_nonVisibleProcessEarlyMemoryCleanupTimer` or `m_nonVisibleProcessMemoryCleanupTimer`?

heycam replied:
Unlike the existing timers, this one needs to be able to run again, later, if the page is still in the background but we draw images. This can happen for example when you move the pointer over the tab strip and we paint the page thumbnails. At that point we should throw away the NativeImages again shortly after. Otherwise it's easy for this patch not to have a real world effect, since the user will probably do this at some point and all the memory usage will come back.
